File: Class/OC/MLScanControl.m
//
//  MLScanControl.m
//  MLScanControl
//
//  Created by Mrlu on 2018/4/23.
//  Copyright © 2018 Mrlu. All rights reserved.
//

#import "MLScanControl.h"
#import <AVFoundation/AVFoundation.h>

@interface MLScanControl()<AVCaptureMetadataOutputObjectsDelegate,AVCaptureVideoDataOutputSampleBufferDelegate>

@property (nonatomic, strong) AVCaptureDevice *device;
@property (nonatomic, strong) AVCaptureDeviceInput *input;
@property (nonatomic, strong) AVCaptureMetadataOutput *output;
@property (nonatomic, strong) AVCaptureVideoDataOutput * videoDataOutput;
@property (nonatomic, strong) AVCaptureSession *scanSession;

@property (nonatomic, strong) AVCaptureVideoPreviewLayer * scanPreviewLayer;
@property (nonatomic, strong) MLScanFrameView * scanFrameView;
@property (nonatomic, strong) UIView * corverLayerView;
@property (nonatomic, strong) CAShapeLayer *maskShapLayer;

@property (nonatomic, strong) UIButton *torchBtn;
@property (nonatomic, assign) BOOL isTurnON;

@property (nonatomic, copy) ResultClosure resultClosure;
@property (nonatomic, assign) CGFloat zoomTemp;

@end

@implementation MLScanControl

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.frameSize = CGSizeMake(260, 260);
        self.style = MLScanStyleWeChat;
        self.offsetY = 0;
        self.zoomTemp = 0;
        self.isTurnON = NO;
        self.isSoundEnable = NO;
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame style:(MLScanStyle)style {
    self = [self initWithFrame:frame];
    if (self) {
        
        self.style = style;
        self.offsetY = -_frameSize.height/2;
    }
    return self;
}

- (void)initialize {
    self.device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    if (self.device) {
        self.scanSession = [[AVCaptureSession alloc] init];
        [self.scanSession canSetSessionPreset:AVCaptureSessionPresetHigh];
        NSError *error = nil;
        self.input = [[AVCaptureDeviceInput alloc] initWithDevice:self.device error:&error];
        if (error) {
            NSLog(@"%@",error);
        }
        self.output = [[AVCaptureMetadataOutput alloc] init];
        [self.output setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
        self.videoDataOutput = [[AVCaptureVideoDataOutput alloc] init];
        [self.videoDataOutput setSampleBufferDelegate:self queue:dispatch_get_main_queue()];
        
        if ([self.scanSession canAddInput:self.input]) {
            [self.scanSession addInput:self.input];
        }
        
        if ([self.scanSession canAddOutput:self.output]) {
            [self.scanSession addOutput:self.output];
        }
        
        if([self.scanSession canAddOutput:self.videoDataOutput]) {
            [self.scanSession addOutput:self.videoDataOutput];
        }
        
        self.output.metadataObjectTypes = @[
                                           AVMetadataObjectTypeQRCode,
                                           AVMetadataObjectTypeCode39Code,
                                           AVMetadataObjectTypeCode93Code,
                                           AVMetadataObjectTypeCode128Code,
                                           AVMetadataObjectTypeCode39Mod43Code,
                                           AVMetadataObjectTypeEAN13Code,
                                           AVMetadataObjectTypeEAN8Code
                                           ];
        
        self.scanPreviewLayer = [[AVCaptureVideoPreviewLayer alloc]initWithSession:self.scanSession];
        self.scanPreviewLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
        self.scanPreviewLayer.frame = self.layer.bounds;
        [self.layer insertSublayer:self.scanPreviewLayer atIndex:0];
        
        [[NSNotificationCenter defaultCenter] addObserverForName:AVCaptureInputPortFormatDescriptionDidChangeNotification object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
            self.output.rectOfInterest = [self.scanPreviewLayer metadataOutputRectOfInterestForRect:self.scanFrameView.frame];
        }];
        
        UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(zoomAction:)];
        tapGesture.numberOfTapsRequired = 2;
        [self addGestureRecognizer:tapGesture];
        
        UIPinchGestureRecognizer *pinchGesture = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(zoomAction:)];
        [self addGestureRecognizer:pinchGesture];
        
    }
}

- (void)setUpView {
    self.backgroundColor = [UIColor blackColor];
    
    self.corverLayerView = [[UIView alloc] initWithFrame:self.bounds];
    self.corverLayerView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
    self.corverLayerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    [self addSubview:self.corverLayerView];
    
    self.scanFrameView = [[MLScanFrameView alloc] initWithFrame:CGRectMake(0, 0, self.frameSize.width, self.frameSize.height) style:self.style];
    self.scanFrameView.center = CGPointMake(self.bounds.size.width/2, self.bounds.size.height/2 + self.offsetY);
    [self addSubview:self.scanFrameView];
    
    self.torchBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    if (self.style == MLScanStyleWeChat) {
        [self.torchBtn setImage:[MLScanControl imageForName:@"weChatScan/ScanLowLight"] forState:UIControlStateNormal];
        [self.torchBtn setImage:[MLScanControl imageForName:@"weChatScan/ScanLowLight_HL"] forState:UIControlStateSelected];
    } else if (self.style == MLScanStyleAlipay) {
        [self.torchBtn setImage:[MLScanControl imageForName:@"zhifuBaoScan/icon_light_off"] forState:UIControlStateNormal];
        [self.torchBtn setImage:[MLScanControl imageForName:@"zhifuBaoScan/icon_light_on"] forState:UIControlStateSelected];
    } else {
        [self.torchBtn setImage:[MLScanControl imageForName:@"weChatScan/ScanLowLight"] forState:UIControlStateNormal];
        [self.torchBtn setImage:[MLScanControl imageForName:@"weChatScan/ScanLowLight_HL"] forState:UIControlStateSelected];
    }
    [self.torchBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.torchBtn.titleLabel.font = [UIFont systemFontOfSize:12];
    self.torchBtn.hidden = YES;
    [self.torchBtn addTarget:self action:@selector(torchBtnAction:) forControlEvents:UIControlEventTouchUpInside];
    self.torchBtn.frame = CGRectMake(0, 0, 40, 40);
    
    self.torchBtn.center = CGPointMake(CGRectGetMidX(self.scanFrameView.frame), CGRectGetMaxY(self.scanFrameView.frame) + 30 + self.torchBtn.frame.size.height/2);
    [self addSubview:self.torchBtn];
}

- (void)didMoveToSuperview {
    [super didMoveToSuperview];
    if (self.superview) {
        [self setUpView];
        __weak typeof(self) weakSelf = self;
        [self authorization:^{
            [weakSelf initialize];
            [weakSelf start];
        } error:^{
            [weakSelf showErrorAlertView];
        }];
    }
}

- (instancetype)result:(ResultClosure)closure {
    self.resultClosure = closure;
    return self;
}

- (void)start {
    if (self.scanSession && !self.scanSession.isRunning) {
        [self.scanSession startRunning];
        [self.scanFrameView startAnimation];
    }
}

- (void)stop {
    if (self.scanSession && self.scanSession.isRunning) {
        [self.scanSession stopRunning];
        [self.scanFrameView stopAnimation];
    }
    self.torchBtn.selected = NO;
}

- (void)authorization:(void(^)(void))authorized error:(void(^)(void))error {
    AVAuthorizationStatus authorizationStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    switch (authorizationStatus) {
        case AVAuthorizationStatusNotDetermined:
        {
            [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        if (granted) {
                            if (authorized) {
                                authorized();
                            }
                        } else {
                            if (error) {
                                error();
                            }
                        }
                    });
            }];
            break;
        }
        case AVAuthorizationStatusAuthorized:
        {
            if (authorized) {
                authorized();
            }
            break;
        }
        default:
        {
            if (error) {
                error();
            }
            break;
        }
    }
}

- (CAShapeLayer *)maskShapLayer {
    if (_maskShapLayer == nil) {
        _maskShapLayer = [[CAShapeLayer alloc] init];
    }
    return _maskShapLayer;
}

- (void)showErrorAlertView {
    UIAlertController *alertVC = [UIAlertController alertControllerWithTitle:@"温馨提示" message:@"请您设置允许该应用访问您的相机" preferredStyle:UIAlertControllerStyleAlert];
    [alertVC addAction:[UIAlertAction actionWithTitle:@"去设置" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString] options:@{} completionHandler:nil];
    }]];
    [[[UIApplication sharedApplication] keyWindow].rootViewController presentViewController:alertVC animated:YES completion:nil];
}

- (void)setTorchBtnEnabel:(BOOL)enabel {
    if (self.torchBtn && !self.torchBtn.selected) {
        self.torchBtn.hidden = !enabel;
        if (!enabel) {
            [self torch:false];
            self.torchBtn.selected = NO;
        }
    }
}

- (void)torchBtnAction:(UIButton *)sender {
    sender.selected = !sender.isSelected;
    [self torch:sender.selected];
}

- (void)zoomAction:(UIGestureRecognizer *)gesture {
    if (self.device) {
        if ([gesture isKindOfClass:[UITapGestureRecognizer class]]) {
            if (self.device.videoZoomFactor == self.device.activeFormat.videoMaxZoomFactor) {
                [self zoom:1 rate:10];
            } else {
                [self zoom:self.device.activeFormat.videoMaxZoomFactor rate:10];
            }
        } else if ([gesture isKindOfClass:[UIPinchGestureRecognizer class]]) {
            if (gesture.state == UIGestureRecognizerStateBegan) {
                self.zoomTemp = self.device.videoZoomFactor;
            }
            if (gesture.state == UIGestureRecognizerStateChanged || gesture.state == UIGestureRecognizerStateEnded) {
                UIPinchGestureRecognizer *pinchGesture = (UIPinchGestureRecognizer *)gesture;
                [self zoom:pinchGesture.scale * self.zoomTemp rate:0];
            }
        }
    }
}

#pragma mark - AVCaptureMetadataOutputObjectsDelegate
- (void)captureOutput:(AVCaptureOutput *)output didOutputMetadataObjects:(NSArray<__kindof AVMetadataObject *> *)metadataObjects fromConnection:(AVCaptureConnection *)connection {
    [self stop];
    [self playSound];
    
    if (metadataObjects.count > 0) {
        if ([metadataObjects.firstObject isKindOfClass:[AVMetadataMachineReadableCodeObject class]]) {
        AVMetadataMachineReadableCodeObject * resultObj = (AVMetadataMachineReadableCodeObject *)metadataObjects.firstObject;
            if (self.resultClosure) {
                self.resultClosure(resultObj.stringValue);
            }
        }
    }
}

#pragma mark - AVCaptureVideoDataOutputSampleBufferDelegate
- (void)captureOutput:(AVCaptureOutput *)output didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection {
    CFDictionaryRef metadataDict = CMCopyDictionaryOfAttachments(kCFAllocatorDefault, sampleBuffer, kCMAttachmentMode_ShouldPropagate);
    NSDictionary *metaData = (__bridge_transfer NSDictionary *)metadataDict;
    NSDictionary *exifMetadata = [[metaData objectForKey:(NSString *)kCGImagePropertyExifDictionary] mutableCopy];
    CGFloat brightnessValue = [[exifMetadata objectForKey:(NSString *)kCGImagePropertyExifBrightnessValue] floatValue];
        if (brightnessValue <= 0) {
            [self setTorchBtnEnabel:YES];
        } else {
            [self setTorchBtnEnabel:NO];
        }
}

#pragma mark - Internal Helpers
- (void)playSound {
    if (!self.isSoundEnable) { return; }
    
    NSString *soundPath = [[NSBundle bundleWithPath:[MLScanControl bundlePath]] pathForResource:@"sound/noticeMusic.caf" ofType:nil];
    NSURL *soundUrl = [NSURL URLWithString:soundPath];
    SystemSoundID soundID = 0;
    AudioServicesCreateSystemSoundID(CFBridgingRetain(soundUrl), &soundID);
}

- (void)torch:(BOOL)isTurnOn {
    if (self.device && self.device.hasTorch) {
        if (isTurnOn) {
            [self.device lockForConfiguration:nil];
            self.device.torchMode = AVCaptureTorchModeOn;
            [self.device unlockForConfiguration];
        } else {
            [self.device lockForConfiguration:nil];
            self.device.torchMode = AVCaptureTorchModeOff;
            [self.device unlockForConfiguration];
        }
    }
}

- (void)zoom:(CGFloat)value rate:(CGFloat)rate {
    if (self.device) {
        [self.device lockForConfiguration:nil];
        if (rate > 0) {
            [self.device rampToVideoZoomFactor:MIN(MAX(value, 1),self.device.activeFormat.videoMaxZoomFactor) withRate:rate];
        } else {
            self.device.videoZoomFactor = MIN(MAX(value, 1),self.device.activeFormat.videoMaxZoomFactor);
        }
        [self.device unlockForConfiguration];
    }
}

+ (NSString *)bundlePath {
    return [[NSBundle mainBundle] pathForResource:@"MLScan" ofType:@"bundle"];
}

+ (UIImage *)imageForName:(NSString *)name {
    NSBundle *bundle = [NSBundle bundleWithPath:[self bundlePath]];
    UIImage *image = [UIImage imageNamed:[NSString stringWithFormat:@"images/%@",name] inBundle:bundle compatibleWithTraitCollection:nil];
    return image;
}

@end

File: Class/Swift/MLScanControl.swift
//
//  MLScanControl.swift
//  SuperLearn_toelf
//
//  Created by Mrlu on 18/01/2018.
//  Copyright © 2018 xdf. All rights reserved.
//

import UIKit
import AVFoundation

class MLScanControl: UIControl {

    typealias ResultClosure = (_ result:String?) -> (Void)
    
    private var device:AVCaptureDevice?
    private var input:AVCaptureDeviceInput?
    private var output:AVCaptureMetadataOutput?
    private var videoDataOutput:AVCaptureVideoDataOutput?
    private var scanSession:AVCaptureSession?
    
    private var scanPreviewLayer:AVCaptureVideoPreviewLayer!
    private var scanFrameView:MLScanFrameView!
    private var corverLayerView:UIView!
    private var maskShapLayer:CAShapeLayer = {
        let maskShapLayer = CAShapeLayer()
        return maskShapLayer
    }()
    private var torchBtn:UIButton?
    private var isTurnON:Bool = false
    var style:MLScanStyle = .weChat
    var frameSize:CGSize = CGSize(width: 260, height: 260)
    var offsetY:CGFloat = 0
    private var resultClosure:ResultClosure?
    private var zoomTemp:CGFloat = 0
    var isSoundEnable:Bool = false
    
    init(frame: CGRect, style:MLScanStyle = .weChat) {
        super.init(frame: frame)
        self.style = style
        offsetY = -frameSize.height/2
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func start() {
        if let scanSession = scanSession, !scanSession.isRunning {
            scanSession.startRunning()
            scanFrameView.startAnimation()
        }
    }
    
    func stop() {
        if let scanSession = scanSession, scanSession.isRunning {
            scanSession.stopRunning()
            scanFrameView.stopAnimation()
        }
        torchBtn?.isSelected = false
    }
    
    @discardableResult
    func result(closure:ResultClosure?) -> Self {
        resultClosure = closure
        return self
    }
    
    private func initialize() {
        //设置捕捉设备
        device = AVCaptureDevice.default(for: AVMediaType.video)
        //设置设备输入输出
        if let device = device {
            
            //设置会话
            scanSession = AVCaptureSession()
            scanSession?.canSetSessionPreset(AVCaptureSession.Preset.high)
            
            do {
                input = try AVCaptureDeviceInput(device: device)
            } catch let error as NSError {
                debugPrint(error)
            }
            output = AVCaptureMetadataOutput()
            output?.setMetadataObjectsDelegate(self, queue: DispatchQueue.main)
            videoDataOutput = AVCaptureVideoDataOutput()
            videoDataOutput?.setSampleBufferDelegate(self, queue: DispatchQueue.main)
            
            if scanSession!.canAddInput(input!) {
                scanSession!.addInput(input!)
            }
            
            if scanSession!.canAddOutput(output!) {
                scanSession!.addOutput(output!)
            }
            
            if scanSession!.canAddOutput(videoDataOutput!) {
                scanSession!.addOutput(videoDataOutput!)
            }
            
            //设置扫描类型(二维码和条形码)
            output?.metadataObjectTypes = [
                AVMetadataObject.ObjectType.qr,
                AVMetadataObject.ObjectType.code39,
                AVMetadataObject.ObjectType.code93,
                AVMetadataObject.ObjectType.code128,
                AVMetadataObject.ObjectType.code39Mod43,
                AVMetadataObject.ObjectType.ean13,
                AVMetadataObject.ObjectType.ean8
                ]
            
            scanPreviewLayer = AVCaptureVideoPreviewLayer(session:scanSession!)
            scanPreviewLayer.videoGravity = AVLayerVideoGravity.resizeAspectFill
            scanPreviewLayer.frame = layer.bounds
            layer.insertSublayer(scanPreviewLayer, at: 0)
            
            //设置扫描区域
            NotificationCenter.default.addObserver(forName: NSNotification.Name.AVCaptureInputPortFormatDescriptionDidChange, object: nil, queue: nil, using: { [weak self] (noti) in
                if let strongSelf = self {
                    strongSelf.output?.rectOfInterest = (strongSelf.scanPreviewLayer?.metadataOutputRectConverted(fromLayerRect: strongSelf.scanFrameView.frame))!
                }
            })
            
            let tapGesture = UITapGestureRecognizer(target: self, action: #selector(zoomAction(sender:)))
            tapGesture.numberOfTapsRequired = 2
            addGestureRecognizer(tapGesture)
            
            let pinchGesture = UIPinchGestureRecognizer(target: self, action: #selector(zoomAction(sender:)))
            addGestureRecognizer(pinchGesture)
        }
    }
    
    private func setUpView() {
        
        backgroundColor = UIColor.black
        
        corverLayerView = UIView(frame: self.bounds)
        corverLayerView.backgroundColor = UIColor.black.withAlphaComponent(0.5)
        corverLayerView.autoresizingMask = [.flexibleHeight,.flexibleWidth]
        addSubview(corverLayerView)
        
        scanFrameView = MLScanFrameView(frame: CGRect(origin: CGPoint.zero, size: frameSize), style:self.style)
        scanFrameView.center = CGPoint(x: bounds.size.width/2, y: bounds.size.height/2 + offsetY)
        addSubview(scanFrameView)
        
        torchBtn = UIButton(type: .custom)
        if style == .weChat {
            torchBtn!.setImage(MLScanControl.image(for: "weChatScan/ScanLowLight"), for: .normal)
            torchBtn!.setImage(MLScanControl.image(for: "weChatScan/ScanLowLight_HL"), for: .selected)
        } else if style == .alipay {
            torchBtn!.setImage(MLScanControl.image(for: "zhifuBaoScan/icon_light_off"), for: .normal)
            torchBtn!.setImage(MLScanControl.image(for: "zhifuBaoScan/icon_light_on"), for: .selected)
        } else {
            torchBtn!.setImage(MLScanControl.image(for: "weChatScan/ScanLowLight"), for: .normal)
            torchBtn!.setImage(MLScanControl.image(for: "weChatScan/ScanLowLight_HL"), for: .selected)
        }
        torchBtn!.setTitleColor(UIColor.white, for: .normal)
        torchBtn!.titleLabel?.font = UIFont.systemFont(ofSize: 12)
//        torchBtn!.setTitle("轻点开启", for: .normal)
//        torchBtn!.setTitle("轻点关闭", for: .selected)
        torchBtn!.isHidden = true
        torchBtn?.addTarget(self, action: #selector(torchBtnAction(sender:)), for: .touchUpInside)
        torchBtn!.frame = CGRect(origin: CGPoint.zero, size:CGSize(width: 40, height: 40))
        torchBtn!.center = CGPoint(x: scanFrameView.frame.midX, y: scanFrameView.frame.maxY + 30 + torchBtn!.frame.height/2)
        addSubview(torchBtn!)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        //mask
        let path = UIBezierPath(rect: self.bounds)
        path.append(UIBezierPath(rect: scanFrameView.frame).reversing())
        self.maskShapLayer.path = path.cgPath
        corverLayerView.layer.mask = self.maskShapLayer
    }
    
    override func didMoveToSuperview() {
        super.didMoveToSuperview()
        if let _ = self.superview {
            setUpView()
            
            authorization(authorized: { [weak self] in
                self?.initialize()
                self?.start()
            }) { [weak self] in
                self?.showErrorAlertView()
            }
        }
    }
    
    private func setTorchBtn(enable:Bool){
        guard let torchBtn = torchBtn, !torchBtn.isSelected else {
            return
        }
        torchBtn.isHidden = !enable
        if !enable {
            torch(isTurnON: false)
            torchBtn.isSelected = false
        }
    }
    
    @objc
    private func torchBtnAction(sender:UIButton) {
        sender.isSelected = !sender.isSelected
        torch(isTurnON: sender.isSelected)
    }
    
    @objc
    private func zoomAction(sender:UIGestureRecognizer) {
        guard let device = device else {
            return
        }
        if sender is UITapGestureRecognizer {
            if device.videoZoomFactor == device.activeFormat.videoMaxZoomFactor {
                zoom(value: 1, rate: 10)
            } else {
                zoom(value: device.activeFormat.videoMaxZoomFactor, rate: 10)
            }
        }
        if let pinchGestureRecognizer = sender as? UIPinchGestureRecognizer {
            if pinchGestureRecognizer.state == UIGestureRecognizerState.began {
                zoomTemp = device.videoZoomFactor
            }
            if pinchGestureRecognizer.state == .changed || pinchGestureRecognizer.state == .ended {
                zoom(value: pinchGestureRecognizer.scale*zoomTemp)
            }
        }
    }
    
    private func authorization(authorized:(()->())?, error:(()->())?) {
        let authorizationStatus = AVCaptureDevice.authorizationStatus(for: AVMediaType.video)
        switch authorizationStatus {
        case.notDetermined:
            AVCaptureDevice.requestAccess(for: AVMediaType.video, completionHandler: { (granted:  Bool) -> Void in
                DispatchQueue.main.async {
                    if granted {
                        authorized?()
                    } else {
                        error?()
                    }
                }
            })
        case.authorized:
            authorized?()
        default:
            error?()
        }
    }
    
    private func showErrorAlertView() {
        let alertVC = UIAlertController(title: "温馨提示", message: "请您设置允许该应用访问您的相机", preferredStyle: UIAlertControllerStyle.alert)
        alertVC.addAction(UIAlertAction(title: "去设置", style:
            UIAlertActionStyle.cancel, handler: { (action) in
                UIApplication.shared.openURL(URL(string: UIApplicationOpenSettingsURLString)!)
        }))
        UIApplication.shared.keyWindow?.rootViewController?.present(alertVC, animated: true, completion: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */

}

extension MLScanControl: AVCaptureMetadataOutputObjectsDelegate {
    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
        //停止扫描
        self.stop()
        playSound()
        
        //扫完完成
        if metadataObjects.count > 0 {
            if let resultObj = metadataObjects.first as? AVMetadataMachineReadableCodeObject {
                // 展示扫描的结果或者其他处理
                resultClosure?(resultObj.stringValue)
            }
        }
    }
}

extension MLScanControl: AVCaptureVideoDataOutputSampleBufferDelegate {
    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        let metadataDict:CFDictionary? = CMCopyDictionaryOfAttachments(kCFAllocatorDefault, sampleBuffer, kCMAttachmentMode_ShouldPropagate)
        let metadata:Dictionary<String,Any>? = metadataDict as? Dictionary<String,Any>
        let exifMetadata:Dictionary<String,Any>? = metadata?[kCGImagePropertyExifDictionary as String] as? Dictionary<String,Any>
        let brightnessValue:Float? = exifMetadata?[kCGImagePropertyExifBrightnessValue as String] as? Float
        if let brightValue = brightnessValue {
            if brightValue <= 0 {
                setTorchBtn(enable: true)
            } else {
                setTorchBtn(enable: false)
            }
        }
    }
}

//MARK - util
extension MLScanControl {
    private func playSound() {
        
        guard self.isSoundEnable else {
            return
        }
        //播放声音
        guard let soundPath = Bundle(path: MLScanControl.bundlePath())?.path(forResource: "sound/noticeMusic.caf", ofType: nil) else { return }
        guard let soundUrl = NSURL(string: soundPath) else { return }
        var soundID:SystemSoundID = 0
        AudioServicesCreateSystemSoundID(soundUrl, &soundID)
        AudioServicesPlaySystemSound(soundID)
    }
    
    private class func bundlePath() -> String {
        return Bundle.main.path(forResource: "MLScan", ofType: "bundle")!
    }
    
    private class func image(for name:String) -> UIImage? {
        let bundle = Bundle(path:Bundle.main.path(forResource: "MLScan", ofType: "bundle")!)
        let image = UIImage(named: "images/" + name, in: bundle, compatibleWith: nil)
        return image
    }
    
    /// torch
    ///
    /// - Parameter isTurnON: isTurnON
    private func torch(isTurnON:Bool) {
        guard let device = device else {
            return
        }
        if device.hasTorch {
            if isTurnON {
                try? device.lockForConfiguration()
                device.torchMode =  AVCaptureDevice.TorchMode.on
                device.unlockForConfiguration()
            } else {
                try? device.lockForConfiguration()
                device.torchMode =  AVCaptureDevice.TorchMode.off
                device.unlockForConfiguration()
            }
        }
    }
    
    /// zoom
    ///
    /// - Parameter value: //1..videoMaxZoomFactor
    private func zoom(value:CGFloat, rate:Float = 0) {
        guard let device = device else {
            return
        }
        try? device.lockForConfiguration()
        if rate > 0 {
            device.ramp(toVideoZoomFactor: min(max(value, 1),device.activeFormat.videoMaxZoomFactor), withRate: rate)
        } else {
            device.videoZoomFactor = min(max(value, 1),device.activeFormat.videoMaxZoomFactor)
        }
        device.unlockForConfiguration()
    }
}
